### ISSUE TITLE
Race Condition in IAM Roles

### ISSUE THREAD

**Issue by gricey432** [Miserlou/Zappa#249]
Sometimes when running a deploy (not update) I get an error `ClientError: An error occurred (InvalidParameterValueException) when calling the CreateFunction operation: The role defined for the function cannot be assumed by Lambda.`

A bit of reading suggests this can be caused by creating an IAM role and then assigning it to something before it's had time to propagate at the AWS end. This seems to make sense because if I wait a minute and then run deploy again it complains the role already exists but then works.

**Comment by Miserlou** [comment]
Dang. Did you find any documentation around that? We might have to poll in a loop in the policy is there. I think this is going to be even worse if/when we move to CF templates..

**Comment by gricey432** [comment]
There's nothing I could find in the AWS docs relating to it, but all the following discussions talk about a race condition.
https://github.com/guardian/cf-notify/issues/1
http://stackoverflow.com/questions/36419442/the-role-defined-for-the-function-cannot-be-assumed-by-lambda
http://stackoverflow.com/questions/37503075/invalidparametervalueexception-the-role-defined-for-the-function-cannot-be-assu

**Comment by Miserlou** [comment]
That hopefully fixes it, anybody who experiences this should feel free to update this ticket if that doesn't work.

**Comment by gene1wood** [comment]
@Miserlou Yes, I just encountered this bug in zappa version 0.40.0. The stacktrace is below. Looks like it happens [here](https://github.com/Miserlou/Zappa/blob/39f05172213fc4ca9dbddb09c908e31d43cc1c36/zappa/zappa.py#L716-L731) which isn't addressed in fa1b224. 

The workaround is to wait a few seconds and re-run deploy.

Should I open a new bug or is this comment good enough?

```
$ venv/bin/zappa deploy dev
...snip...

==============

Traceback (most recent call last):
  File "/path/to/zappa/venv/local/lib/python2.7/site-packages/zappa/cli.py", line 2187, in handle
    sys.exit(cli.handle())
  File "/path/to/zappa/venv/local/lib/python2.7/site-packages/zappa/cli.py", line 414, in handle
    self.dispatch_command(self.command, environment)
  File "/path/to/zappa/venv/local/lib/python2.7/site-packages/zappa/cli.py", line 446, in dispatch_command
    self.deploy()
  File "/path/to/zappa/venv/local/lib/python2.7/site-packages/zappa/cli.py", line 607, in deploy
    memory_size=self.memory_size
  File "/path/to/zappa/venv/local/lib/python2.7/site-packages/zappa/zappa.py", line 730, in create_lambda_function
    DeadLetterConfig=dead_letter_config
  File "/path/to/zappa/venv/local/lib/python2.7/site-packages/botocore-1.5.16-py2.7.egg/botocore/client.py", line 253, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/path/to/zappa/venv/local/lib/python2.7/site-packages/botocore-1.5.16-py2.7.egg/botocore/client.py", line 543, in _make_api_call
    raise error_class(parsed_response, operation_name)
InvalidParameterValueException: An error occurred (InvalidParameterValueException) when calling the CreateFunction operation: The role defined for the function cannot be assumed by Lambda.

==============

Need help? Found a bug? Let us know! :D
```

**Comment by gricey432** [comment]
A 7 second delay seems to be fine most of the time but could still fail if aws is under load or there are propagation issues

**Comment by adamdavis40208** [comment]
I ran into it on a fresh copy of 0.42.0, on a simple zappa init/zappa deploy dev.

Running zappa deploy a second time did not run into the issue though

**Comment by gene1wood** [comment]
I'm assuming commenting in this closed issue is probably not ideal so I opened #901 since this bug is still happening.

**Comment by akravetz** [comment]
I am experiencing this issue as well.  Even after trying "zappa deploy" again the error remains the same

**Comment by gene1wood** [comment]
@akravetz as this bug is closed you may want to follow #901
